PC/launcher.c

```diff
@@ -1040,32 +1046,47 @@ find_terminator(char * buffer, int len, BOM *bom)
 static BOOL
 validate_version(wchar_t * p)
 {
-    BOOL result = TRUE;
-
-    if (!isdigit(*p))               /* expect major version */
-        result = FALSE;
-    else if (*++p) {                /* more to do */
-        if (*p != L'.')             /* major/minor separator */
+    /*
+    Version information should start with one of 2 or 3,
+    Optionally followed by a period and a minor version,
+    Optionally followed by a minus and one of 32 or 64.
+    Valid examples:
+      2
+      3
+      2.7
+      3.6
+      2.7-32
+      The intent is to add to the valid patterns:
+      3.10
+      3-32
+      3.6-64
+      3-64
+    */
+    BOOL result = (p != NULL) && *p;  /* Default to false if empty string or null pointer. */
+
+    while (result && isdigit(*p))   /* Require a major version */{
+        ++p;  /* Skip leading digit(s) */}
+    if (result && (*p == L'.'))     /* Allow . for major minor separator.*/
+    {
+        result = isdigit(*++p);     /* Must be at least one digit */
+        while (isdigit(*++p)) ;     /* Skip an more Digits */
```

Comment:
What if `*++p` in the previous statement on 1071 was the trailing null and `result` is false, i.e. the version string is something like "2."? Now you're reading past the end of the string.